### PR TITLE
Error if id(from file) and supplied proposal name do not match

### PR DESCRIPTION
**Pull request by sjamgade**
while creating a proposal from file

### CODE REVIEW

**Comment by sjamgade** [comment]
from internal [trello](https://trello.com/c/viQGsh8I/268-p2-bsc-1037374-l3-crowbarctl-create-proposal-from-file-always-fails)

**Comment by sjamgade** [comment]
@rsalevsky added the tkt number to commit message. 

```
root@socadmin:/tmp # crowbarctl proposal create crowbar default --file=crowbar-default-proposal
{"error":"Please limit name to [A-z0-9_] without spaces."}
```

**Comment by dirkmueller** [comment]
so the file name needs to be renamed to "default" and then it works? is that what the check is trying to achieve?

**Comment by sjamgade** [comment]
`crowbarctl proposal create crowbar default --file=crowbar.default.proposal`
while creating proposal from a file, the name is taken from `id` field in the file, and the one provided on the cmdline is ignored. 

The file 'crowbar-default-proposal' is a template obtained from the server. These templates follow the pattern `{bcname}-default` for the `id` field. However the `-` in the name is not allowed, as the error points out. But from the users point of view, name is what is provided on the cmdline.

The fix is to make sure `err if id-from-file != name-on-cmdline`

**Comment by sjamgade** [comment]
I hope this helps
```
root@crowbar:~ # grep id crowb                                                                                        
  "id": "crowbar-default",
root@crowbar:~ # crowbarctl proposal create crowbar name-with-hyphen  --file=crowb                                    
{"error":"Please limit name to [A-z0-9_] without spaces."}
root@crowbar:~ # sed -i 's:crowbar-default:crowbardefault:' crowb
root@crowbar:~ # grep id crowb
  "id": "crowbardefault",
root@crowbar:~ # crowbarctl proposal create crowbar name-with-hyphen  --file=crowb 
Successfully created name-with-hyphen proposal
root@crowbar:~ # crowbarctl proposal list crowbar 
+----------------+
| Proposal       |
+----------------+
| crowbardefault |
| default        |
+----------------+
```
-------------------
As @rsalevsky  pointed out, better flow would be to make the `name` optional, and use it from the file if not provided

File: lib/crowbar/client/command/proposal/create.rb [hunk 1]
```diff
@@ -103,6 +103,10 @@ def from_file
             rescue
               err "Failed to process file"
             end
+            args.proposal = args.proposal || json["id"]
+            if json["id"] != args.proposal
+              err "id(#{json["id"]}) in file and name(#{args.proposal}) should be same"
```
Comment (rsalevsky) [inline lib/crowbar/client/command/proposal/create.rb]: In case the id and the arg is not the same I would just use the id and give a warning.

File: lib/crowbar/client/app/proposal.rb [hunk 2]
```diff
@@ -141,7 +141,7 @@ def show(barclamp, proposal)
           catch_errors(e)
         end
 
-        desc "create BARCLAMP PROPOSAL",
+        desc "create BARCLAMP [PROPOSAL]",
```
Comment (rsalevsky) [inline lib/crowbar/client/app/proposal.rb]: We should do the same for all other barclamp commands.

**Comment by rsalevsky** [comment]
I would suggest to do the change in one commit also pls use the default schema for commit messages. In this case it would be `proposal: ....... (bsc#XXXXX)`.

File: lib/crowbar/client/app/proposal.rb (hunk 2 above)
Comment (sjamgade) [inline lib/crowbar/client/app/proposal.rb]: May be we should get that in different PR.

Comment (rsalevsky) [inline lib/crowbar/client/app/proposal.rb]: Works for me.

File: lib/crowbar/client/command/proposal/create.rb [hunk 3]
```diff
@@ -103,6 +103,11 @@ def from_file
             rescue
               err "Failed to process file"
             end
+            if json["id"] != args.proposal
```
Comment (rsalevsky) [inline lib/crowbar/client/command/proposal/create.rb]: Can you add a empty line above? Sorry to be picky. :P

File: lib/crowbar/client/command/proposal/create.rb [hunk 4]
```diff
@@ -103,6 +103,11 @@ def from_file
...
+              say "id(#{json["id"]}) in file and name(#{args.proposal}) do not match"
```
Comment (rsalevsky) [inline lib/crowbar/client/command/proposal/create.rb]: I'm not sure but maybe the message is a bit too much cryptic. What about: `Given proposal name and file proposal name do not match, using #{json["id"]}.`? Then you could also drop the second `say`.

File: lib/crowbar/client/command/proposal/create.rb (hunk 3 above)
Comment (sjamgade) [inline lib/crowbar/client/command/proposal/create.rb]: its okay :)

File: lib/crowbar/client/command/proposal/create.rb (hunk 4 above)
Comment (sjamgade) [inline lib/crowbar/client/command/proposal/create.rb]: Its okay. Done.

File: lib/crowbar/client/command/proposal/create.rb [hunk 5]
```diff
@@ -104,6 +104,11 @@ def from_file
               err "Failed to process file"
             end
 
+            if json["id"] != args.proposal
+              args.proposal = json["id"]
```
Comment (dirkmueller) [inline lib/crowbar/client/command/proposal/create.rb]: it is possible to assign to cmdline args? this sounds weird to me

File: lib/crowbar/client/command/proposal/create.rb [hunk 6]
```diff
@@ -104,6 +104,11 @@ def from_file
...
+              say "Given proposal name and file proposal name do not match, using id(#{json["id"]})"
```
Comment (dirkmueller) [inline lib/crowbar/client/command/proposal/create.rb]: maybe simplify to

```ruby

   say "Using #{json["id"]} as proposal name"
```

since there no way to not do that anymore? what about options[:merge] ? should that be disabled?

File: lib/crowbar/client/app/proposal.rb [hunk 7]
```diff
@@ -189,7 +191,7 @@ def show(barclamp, proposal)
           aliases: ["-m"],
           desc: "Merge input loaded from server with proposal data"
 
-        def create(barclamp, proposal)
+        def create(barclamp, proposal = "default")
```
Comment (dirkmueller) [inline lib/crowbar/client/app/proposal.rb]: should that default be set on args instead?

**Comment by dirkmueller** [comment]
speaking without knowing any background of this, I would expect this behavior

> $ crowbar proposal create --file=foo.yaml
* creates a proposal with the name "default"
* errors out if a proposal with name "default" already existsw
* overwrites the id in the json with the proposal id "default" if not matching (or alternatively errors out)

> $ crowbar proposal create --file=foo.yaml somethingelse
*  creates a proposal with the name "somethingelse"
* if foo.yaml has an id "something", it overwrites it with "somethingelse"
